File: 2023/day07/perl/solution.pl
#!/usr/bin/env perl
use v5.16;
use strict;
use warnings;
use File::Basename;
use File::Spec;
use List::Util qw(sum);

# Read input file
my $dir = dirname(__FILE__);
my $input_path = File::Spec->catfile($dir, '..', 'input.txt');
open(my $fh, '<', $input_path) or die "Cannot open $input_path: $!";
my @lines = <$fh>;
close($fh);
chomp(@lines);

# Card strength order (higher index = stronger)
my $CARD_STRENGTH = "23456789TJQKA";
my $CARD_STRENGTH_JOKER = "J23456789TQKA";  # J is weakest in Part 2

# Pattern-to-type mapping (shared between both hand type functions)
my %PATTERN_TYPE = (
    "5"       => 6,  # Five of a kind
    "4,1"     => 5,  # Four of a kind
    "3,2"     => 4,  # Full house
    "3,1,1"   => 3,  # Three of a kind
    "2,2,1"   => 2,  # Two pair
    "2,1,1,1" => 1,  # One pair
    "1,1,1,1,1" => 0,  # High card
);

sub pattern_to_type {
    my ($pattern) = @_;
    return $PATTERN_TYPE{$pattern} // 0;
}

sub get_hand_type {
    my ($hand) = @_;

    # Count cards
    my %counts;
    $counts{$_}++ for split //, $hand;

    # Sort counts in descending order and convert to pattern
    my $pattern = join(",", sort { $b <=> $a } values %counts);

    return pattern_to_type($pattern);
}

sub get_hand_type_with_jokers {
    my ($hand) = @_;

    my $joker_count = ($hand =~ tr/J//);

    return get_hand_type($hand) if $joker_count == 0;
    return 6 if $joker_count == 5;  # Five of a kind

    # Count non-joker cards
    my %counts;
    for my $c (split //, $hand) {
        $counts{$c}++ if $c ne 'J';
    }

    # Sort counts in descending order
    my @sorted_counts = sort { $b <=> $a } values %counts;

    # Add jokers to the highest count
    $sorted_counts[0] += $joker_count;

    # Convert to pattern and look up type
    my $pattern = join(",", @sorted_counts);

    return pattern_to_type($pattern);
}

sub hand_key {
    my ($hand, $use_jokers) = @_;

    my $hand_type = $use_jokers ? get_hand_type_with_jokers($hand) : get_hand_type($hand);
    my $strength = $use_jokers ? $CARD_STRENGTH_JOKER : $CARD_STRENGTH;
    my @card_values = map { index($strength, $_) } split //, $hand;

    return ($hand_type, @card_values);
}

sub compare_hands {
    my ($a_hand, $b_hand, $use_jokers) = @_;

    my @a_key = hand_key($a_hand, $use_jokers);
    my @b_key = hand_key($b_hand, $use_jokers);

    for my $i (0 .. $#a_key) {
        return $a_key[$i] <=> $b_key[$i] if $a_key[$i] != $b_key[$i];
    }
    return 0;
}

sub solve {
    my ($use_jokers) = @_;

    my @hands;
    for my $line (@lines) {
        next unless $line =~ /\S/;
        my ($hand, $bid) = split /\s+/, $line;
        push @hands, [$hand, $bid];
    }

    # Sort by hand strength
    my @sorted = sort { compare_hands($a->[0], $b->[0], $use_jokers) } @hands;

    # Calculate total winnings using sum from List::Util
    return sum map { ($_ + 1) * $sorted[$_][1] } 0 .. $#sorted;
}

sub part1 { solve(0) }
sub part2 { solve(1) }

say "Part 1: " . part1();
say "Part 2: " . part2();

File: 2023/day07/rust/solution.rs
use std::cmp::Ordering;
use std::collections::HashMap;
use std::fs;

fn classify_hand_type(count_values: &[u8]) -> u8 {
    match count_values {
        [5] => 6,           // Five of a kind
        [4, 1] => 5,        // Four of a kind
        [3, 2] => 4,        // Full house
        [3, 1, 1] => 3,     // Three of a kind
        [2, 2, 1] => 2,     // Two pair
        [2, 1, 1, 1] => 1,  // One pair
        _ => 0,             // High card
    }
}

fn get_hand_type(hand: &str) -> u8 {
    let mut counts: HashMap<char, u8> = HashMap::new();
    for c in hand.chars() {
        *counts.entry(c).or_insert(0) += 1;
    }

    let mut count_values: Vec<u8> = counts.values().copied().collect();
    count_values.sort_unstable_by(|a, b| b.cmp(a));

    classify_hand_type(&count_values)
}

fn get_hand_type_with_jokers(hand: &str) -> u8 {
    let joker_count = hand.chars().filter(|&c| c == 'J').count() as u8;

    if joker_count == 0 {
        return get_hand_type(hand);
    }
    if joker_count == 5 {
        return 6; // Five of a kind
    }

    // Count non-joker cards
    let mut counts: HashMap<char, u8> = HashMap::new();
    for c in hand.chars() {
        if c != 'J' {
            *counts.entry(c).or_insert(0) += 1;
        }
    }

    let mut count_values: Vec<u8> = counts.values().copied().collect();
    count_values.sort_unstable_by(|a, b| b.cmp(a));

    // Add jokers to the highest count
    count_values[0] += joker_count;

    classify_hand_type(&count_values)
}

fn card_strength(c: char) -> u8 {
    const STRENGTH: &str = "23456789TJQKA";
    STRENGTH.find(c).expect("invalid card character") as u8
}

fn card_strength_joker(c: char) -> u8 {
    const STRENGTH: &str = "J23456789TQKA";
    STRENGTH.find(c).expect("invalid card character") as u8
}

fn compare_hands(a: &str, b: &str) -> Ordering {
    let type_a = get_hand_type(a);
    let type_b = get_hand_type(b);

    match type_a.cmp(&type_b) {
        Ordering::Equal => {
            // Compare card by card
            for (ca, cb) in a.chars().zip(b.chars()) {
                let strength_cmp = card_strength(ca).cmp(&card_strength(cb));
                if strength_cmp != Ordering::Equal {
                    return strength_cmp;
                }
            }
            Ordering::Equal
        }
        other => other,
    }
}

fn compare_hands_joker(a: &str, b: &str) -> Ordering {
    let type_a = get_hand_type_with_jokers(a);
    let type_b = get_hand_type_with_jokers(b);

    match type_a.cmp(&type_b) {
        Ordering::Equal => {
            // Compare card by card
            for (ca, cb) in a.chars().zip(b.chars()) {
                let strength_cmp = card_strength_joker(ca).cmp(&card_strength_joker(cb));
                if strength_cmp != Ordering::Equal {
                    return strength_cmp;
                }
            }
            Ordering::Equal
        }
        other => other,
    }
}

fn part1(hands: &[(String, u64)]) -> u64 {
    let mut sorted_hands: Vec<_> = hands.iter().collect();
    sorted_hands.sort_unstable_by(|a, b| compare_hands(&a.0, &b.0));

    sorted_hands
        .iter()
        .enumerate()
        .map(|(i, (_, bid))| (i as u64 + 1) * bid)
        .sum()
}

fn part2(hands: &[(String, u64)]) -> u64 {
    let mut sorted_hands: Vec<_> = hands.iter().collect();
    sorted_hands.sort_unstable_by(|a, b| compare_hands_joker(&a.0, &b.0));

    sorted_hands
        .iter()
        .enumerate()
        .map(|(i, (_, bid))| (i as u64 + 1) * bid)
        .sum()
}

fn main() {
    let input = fs::read_to_string("../input.txt").expect("Failed to read input file");

    let hands: Vec<(String, u64)> = input
        .trim()
        .lines()
        .map(|line| {
            let (hand, bid) = line.split_once(' ').expect("invalid line format: missing space");
            (hand.to_string(), bid.parse().expect("invalid bid: not a number"))
        })
        .collect();

    println!("Part 1: {}", part1(&hands));
    println!("Part 2: {}", part2(&hands));
}
